sql: improve observability of index merge timestamp

spilchen opened the pull request:
When building an index, the merge phase copies any new rows added since the index build started. A timestamp is used to set an upper bound for this scan. We suspect that the chosen timestamp may be omitting some rows.

Our theory is that this happens if new rows were inserted before the merge phase but were committed with a timestamp after the merge. This discrepancy could be due to clock skew between nodes. However, attempts to reproduce this issue via unit tests were unsuccessful, so this remains a hypothesis.

To address this, this change includes:
- An adjustment to the merge timestamp to account for potential clock skew.
- Additional logging of the merge timestamp chosen and the timestamps observed on each node when draining leased descriptors. These logs will help track the timestamp across nodes during the merge phase.
- Treat index validation errors for non-unique indexes as assertion failures. We previously treated as a duplicate key error, which was very confusing because duplicates are allowed for non-unique indexes.

Epic: none
Release note: none
Closes: #142861 
Closes #143050

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142917)
<!-- Reviewable:end -->

fqazi reviewed (changes requested):
Looks fairly good from a diagnostic view point / potential fix. Some suggestions for the unit test

> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142917#-:-OLL0bEj8QH4RYxDb0iY:bxeug59)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/schemachanger/scexec/exec_backfill_test.go` line 395 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142917#-OLL0M6H5R9SgjPZ80zo:-OLKwmrECWpx84lBEbgm:b-v1wmav) ([raw file](https://github.com/cockroachdb/cockroach/blob/c22111bea93afdeb3e843e8eda48cac9b08000e7/pkg/sql/schemachanger/scexec/exec_backfill_test.go#L395)):*
> ```Go
> 			keyVal := 1000 + additionalInserts
> 			additionalInserts++
> 			r1.Exec(t, fmt.Sprintf("INSERT INTO t_idx(n) VALUES (%d)", keyVal))
> ```

This only guarantees the gateway node not where the insert will hit. We might need a zone config tweak to control this so that placement is intentionally skewed in some way. Maybe modify the default range or something goofy.
___
*[`pkg/sql/backfill.go` line 3007 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142917#-OLKrNFoA-6bLCjh53Oc:-OLKrNFoA-6bLCjh53Od:b-e9zza0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c22111bea93afdeb3e843e8eda48cac9b08000e7/pkg/sql/backfill.go#L3007)):*
> ```Go
> 	// that are already present in the index being merged, but that’s fine as
> 	// they will be treated as no-ops.
> 	ts := clock.Now().AddDuration(clock.MaxOffset())
> ```

FYI:
Not an issue, but this should be safe because the txn will go through maybeCommitWait, which will wait for this timestamp if this time is in the future. I think the reason we would into trouble in the past is that the global uncertainty limit doesn't apply to fixed timestamp txns.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql: improve observability of index merge timestamp

When building an index, the merge phase copies any new rows added since
the index build started. A timestamp is used to set an upper bound for
this scan. We suspect that the chosen timestamp may be omitting some
rows.

Our theory is that this happens if new rows were inserted before the
merge phase but were committed with a timestamp after the merge. This
discrepancy could be due to clock skew between nodes. However, attempts
to reproduce this issue via unit tests were unsuccessful, so this
remains a hypothesis.

To address this, this change includes:
- An adjustment to the merge timestamp to account for potential clock
skew.
- Additional logging of the merge timestamp chosen and the timestamps
observed on each node when draining leased descriptors. These logs will
help track the timestamp across nodes during the merge phase.
- Treat index validation errors for non-unique indexes as assertion
failures. We previously treated as a duplicat key error, which was
very confusing because duplicates are allowed for non-unique indexes.

Epic: none
Release note: none
Closes: #142751

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142917)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/schemachanger/scexec/exec_backfill_test.go` line 395 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142917#-OLL0M6H5R9SgjPZ80zo:-OLLycMO3Lm3O3KeinmK:b-fsxeik) ([raw file](https://github.com/cockroachdb/cockroach/blob/c22111bea93afdeb3e843e8eda48cac9b08000e7/pkg/sql/schemachanger/scexec/exec_backfill_test.go#L395)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

This only guarantees the gateway node not where the insert will hit. We might need a zone config tweak to control this so that placement is intentionally skewed in some way. Maybe modify the default range or something goofy.
</blockquote></details>

I tried modifying the zone config so that rows only exist at one node. But I still didn't have any luck reproducing the issue.


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
The changes here are fine, we can maybe try them on DRT to confirm if it helps.
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142917#-:-OL_W8rH5y7zNDxFeSCN:b-9kliv4)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/schemachanger/scexec/exec_backfill_test.go` line 395 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142917#-OLL0M6H5R9SgjPZ80zo:-OL_W0td2M1cXypjUr_5:b-ar16cf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c22111bea93afdeb3e843e8eda48cac9b08000e7/pkg/sql/schemachanger/scexec/exec_backfill_test.go#L395)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

I tried modifying the zone config so that rows only exist at one node. But I still didn't have any luck reproducing the issue.
</blockquote></details>

Does executing on r1 make a difference?


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142917)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/schemachanger/scexec/exec_backfill_test.go` line 395 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142917#-OLL0M6H5R9SgjPZ80zo:-OLcuMu3EpQ_0fmuDzWZ:b3t4m6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c22111bea93afdeb3e843e8eda48cac9b08000e7/pkg/sql/schemachanger/scexec/exec_backfill_test.go#L395)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Does executing on r1 make a difference?
</blockquote></details>

No, it doesn't. I think I tried every possible combination.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
